### DISCUSSION TITLE
How To Mock An External Package Sub Folder File

### DISCUSSION THREAD

**Discussion by DevYemi** [vitest-dev/vitest#5119] [Q&A] [answered]
Hi there guys,

So i have gone through vitest docs covering modules mocking, the example in the docs only illustrates how to mock a default export I think which is the `axios.js` example here [Here](https://vitest.dev/api/vi.html#mock-modules).

But I am trying to mock `next-auth/react` package and I am kinda confused about how to go about it.

what I have tried is to create a `__mocks__` folder at the root of my project as the docs suggested and added a folder `next-auth/react.js`

> `__mocks__`
>> `next-auth`
>>> `react.js`

![Screenshot 2024-02-05 233055](https://github.com/vitest-dev/vitest/assets/68167320/8dd045f8-a9ed-4617-8d6b-0d832d9ad6a2)

Now in the `react.js` file I added the code where I am mocking the exported function I want to mock.

````
import { vi } from "vitest";
import * as nextAuthReactModule from 'next-auth/react'



export const modObject = {
    ...nextAuthReactModule,
    useSession: vi.fn(),
    getSession: vi.fn(),
} 
````

Then in my `setupTest.ts` file I am doing `vi.mock("next-auth/react");` to mock the module before each test runs.

But it's not working, seems like vitest is not catching the mocks I created and it's still using the original module when my code runs which makes my test fail.
 Any help will be appreciated as to what i might be doing wrong.

**Comment by hi-ogawa** [comment]
We have tests using `vi.mock('@vueuse/integrations/useJwt')` with `__mocks__/@vueuse/integrations/useJwt.ts`, so I don't think sub folders is an issue.

https://github.com/vitest-dev/vitest/blob/fec9ca0b9c5356648d821dffd515dc9fb217c531/examples/mocks/test/integration.test.ts#L3

https://github.com/vitest-dev/vitest/blob/fec9ca0b9c5356648d821dffd515dc9fb217c531/examples/mocks/__mocks__/%40vueuse/integrations/useJwt.ts#L3

Can you provide a reproduction to show how you are mocking `next-auth/react` currently?
You can find templates to start with here https://github.com/vitest-dev/vitest/tree/main/examples

**Reply by DevYemi** [reply]
Hi @hi-ogawa thank you so much for responding, i went through the repo examples you suggested but the nextJs example there uses the Page Router while I am using App Router in my project so I decided to create a minimal reproduction of my issue [Link to Github repo ](https://github.com/DevYemi/vitest-mock-issue-minimal-reproduction)

Thank you

**Reply by DevYemi** [reply]
@hi-ogawa, just a side note I was able to make the test pass by mocking the `next-auth/react` like this in the test file but the `__mocks__` convention still fails

````
vi.mock("next-auth/react", async (importOriginal) => {
    const mod = await importOriginal<typeof import('next-auth/react')>();

    const mockSession: Session | null = {
        userData: {
            id: "",
            type: "",
            firstName: "Adeyemi",
            lastName: "Adeyanju",
            email: "dummy.email007@gmail.com",
            phoneNo: "+7383837882",
            avatar: null,
        },
        user: undefined,
        expires: new Date(Date.now() + 2 * 86400).toISOString(),
    };

    const mockData = { data: mockSession, status: 'authenticated' }

    return {
        ...mod,
        useSession: vi.fn(() => mockData),
        getSession: vi.fn(() => mockData),
    }
});
````

**Reply by hi-ogawa** [reply, marked as answer]
Thanks for the reproduction. I think you would need `export const useSession = ...` in  `__mocks__/next-auto/react.js`:

https://stackblitz.com/edit/github-ttporm?file=__mocks__%2Fnext-auth%2Freact.js

I see this might be cumbersome to do for each export, but if that's the case, then using `vi.mock` with `importOriginal` would make sense like you already do.

**Reply by DevYemi** [reply]
Thank you @hi-ogawa, works now thanks to your insights
